kuwala/core/canvas/src/data/state/models/CanvasModel.js

```diff
@@ -0,0 +1,333 @@
+        let isDuplicate = false;
+        state.canvasSelectedDataSource.map((el) => {
+            if (el.id === selectedDataSource.id) isDuplicate = true
+        });
+
+        if (!isDuplicate){
+            state.canvasSelectedDataSource = [...state.canvasSelectedDataSource, selectedDataSource]
+        }
+
+    }),
+
+    // Selected Column Address Action
+    addSelectedColumnAddress: action((state, newAddress) => {
+        let isDuplicate = false;
+        const addressArray = newAddress.split('@');
+        const schema = addressArray[0];
+        const category = addressArray[1];
+        const table = addressArray[2];
+        const column = addressArray[3];
+
+        let tempState = state.selectedAddressObj;
+        if(typeof tempState[schema] === 'undefined') {
+            tempState[schema] = {
+                [category]: {
+                    [table]: [column]
+                }
+            }
+        } else if (typeof tempState[schema][category] === 'undefined'){
+            tempState[schema][category] = {
+                    [table]: [column]
+            }
+        } else if (typeof tempState[schema][category][table] === 'undefined') {
+            tempState[schema][category][table] = [column]
+        } else {
+            tempState[schema][category][table].push(column)
+            tempState[schema][category][table].forEach((el) => {
+                if (el === newAddress) isDuplicate = true
+            });
+
+            if (!isDuplicate){
+                state.selectedAddressObj = tempState
+            }
+        }
+    }),
+
+    removeSelectedColumnAddress: action((state, addressToRemove) => {
+        const {schema, category, table, column} = columnAddressSplitter(addressToRemove);
+        let tempState = state.selectedAddressObj
+        try {
+            const newSelectedTableList = tempState[schema][category][table].filter((el) => el !== column)
+            tempState[schema][category][table] = newSelectedTableList
+        } catch (e) {
+            console.error(e)
+        }
+        state.selectedAddressObj = tempState
+    }),
+
+    generateStructureIfNotExists: action((state, columnAdress) => {
+        const {schema, category, table} = columnAddressSplitter(columnAdress);
+
+        let tempState = state.selectedAddressObj;
+        if(typeof tempState[schema] === 'undefined') {
+            tempState[schema] = {
+                [category]: {
+                    [table]: []
+                }
+            }
+        } else if (typeof tempState[schema][category] === 'undefined'){
+            tempState[schema][category] = {
+                [table]: []
+            }
+        } else if (typeof tempState[schema][category][table] === 'undefined') {
+            tempState[schema][category][table] = []
+        }
+
+        state.selectedAddressObj = tempState
+    }),
+
+    selectAllColumnAddresses: thunk((actions, bulkAddress) => {
+        if(bulkAddress.length <= 0 || !bulkAddress) return
```

Comment:
Should be `if (!bulkAddress || !bulkAddress.length)` since if `bulkAddress` is undefined the length check will throw an error and the length check syntax can be shortened.